### ISSUE TITLE
Tacotron 2: Exploding Loss

### ISSUE THREAD

**Issue by knbunmlr** [NVIDIA/NeMo#282]
Hey guys,

first of all: thanks for the great work!
Then: I was training Tacotron2 with the provided /examples/tts/tacotron2.py on my own dataset. 
The first ~13k steps everything seemed alright but then the loss exploded, cf. the attached pictures.
Any ideas about what could be wrong? In which direction should I investigate? Did anyone observe similar problems?

More information: 
First warning in logs:
2020-01-17 05:41:46,084 - INFO - Step: 13550
2020-01-17 05:41:46,088 - INFO - Step time: 3.824190855026245 seconds
2020-01-17 05:41:57,074 - INFO - Finished epoch 250 in 198.93912196159363
2020-01-17 05:41:57,077 - INFO - Starting epoch 251
2020-01-17 05:43:17,592 - WARNING - WARNING: Loss is NaN

Then training is continued with increasing warning rate until almost only Warnings appear.

**Edit:** Training with O0 solves this issue.

![image](https://user-images.githubusercontent.com/60105425/72738308-955e9a80-3ba1-11ea-9e0a-1d382005821f.png)
![image](https://user-images.githubusercontent.com/60105425/72738359-adceb500-3ba1-11ea-9281-6a47901cf03f.png)
![image](https://user-images.githubusercontent.com/60105425/72738450-e2db0780-3ba1-11ea-9d18-47eb1b009dee.png)
![image](https://user-images.githubusercontent.com/60105425/72738468-ed959c80-3ba1-11ea-89b7-b771839e3d2f.png)

**Comment by knbunmlr** [comment]
Just an idea:

I accidentally started the training with max_step a magnitude higher than the defaults. This of course impacts the Cosine annealing learning policy, i.e. the learning rate doesn't decay as fast as it would with the original max_step, cf.:

# train model
    total_steps = (args.max_steps if args.max_steps is not None else
                   args.num_epochs * steps_per_epoch)
    neural_factory.train(
        tensors_to_optimize=[train_loss],
        callbacks=callbacks,
        lr_policy=CosineAnnealing(total_steps, min_lr=args.min_lr),
        optimizer=args.optimizer,
        optimization_params={
            "num_epochs": args.num_epochs,
            "max_steps": args.max_steps,
            "lr": args.lr,
            "weight_decay": args.weight_decay,
            "grad_norm_clip": args.grad_norm_clip},
        batches_per_step=args.iter_per_step)

Could this be a problem?

**Comment by blisc** [comment]
A couple observations:
- I'm surprised that despite your loss exploding, your attention still seems stable.
- Looking at your attention plots, it seems like your first 100 steps actually doesn't learn any attention, does your audio have some sort of noise or silence at the beginning of the recordings?
- It also seems like your attention plots are clipped, they always seem to reach the end of the decoder timesteps. Are you picking plots that max out the decoder timesteps or is this a consistent behaviour? 
- What happens if you use a constant 1e-3 learning rate

**Comment by knbunmlr** [comment]
Thanks for taking the time.
I'll attach the complete tensorboard file for a more complete picture. Especially interesting: The training mel and gate predictions look actually quite good, whereas the test equivalents are really bad.

I'm using parts of the MAILABS dataset which is sampled with 16kHz. Is setting the sr in the yaml enough to coupe with that change? Maybe a change alà the one from openseq2seq is necessary:
'''
if dataset == "MAILABS":
  trim = True
  mag_num_feats = 401
'''
Also,  this dataset has silence in the beginning and end of the files. I hoped to deal with this fact by adding 'trim_silence: true' to the AudioToTextDataLayer in the config.yaml, cf. attached file.

With the new information, do you still think retraining with a fixed learning rate adds insight? Keep in mind that with the high max_step size there was almost no change in the lr. 
[events.out.tfevents.1579188881.4fc362650846.zip](https://github.com/NVIDIA/NeMo/files/4093609/events.out.tfevents.1579188881.4fc362650846.zip)
[tacotron2-eva_2020-01-16_15-34-41.yaml.zip](https://github.com/NVIDIA/NeMo/files/4093619/tacotron2-eva_2020-01-16_15-34-41.yaml.zip)

**Comment by blisc** [comment]
Have you cleaned your transcripts to only contain these characters:
[' ', '!', "'", ',', '-', '.', ':', ';', '?', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', 'ß', 'ä', 'ö', 'ü'] ?

Looking at your tensorboard logs, it seems like attention collapsed around 15k iterations.
Generally, training mel will always look good since we are using teacher forcing.

**Comment by knbunmlr** [comment]
> Have you cleaned your transcripts to only contain these characters:
[' ', '!', "'", ',', '-', '.', ':', ';', '?', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', 'ß', 'ä', 'ö', 'ü'] ?

Jep, I checked it three times, so I'm quite sure that's correct. Btw, do you think keeping capital letters improves performance?

In the meantime I restarted training with max_steps as 30000, cf. https://nvidia.github.io/NeMo/tts/tutorial.html, but then no attention is learned at all, see the attached tensorboard. [events.out.tfevents.1579536350.22c084f99cbc.zip](https://github.com/NVIDIA/NeMo/files/4096277/events.out.tfevents.1579536350.22c084f99cbc.zip)

Funny thing is, the same dataset trains the taco2 implementation of openseq2seq quite smoothly. Is there a reason why the learning rate policy has changed since then? There we trained for the first 45k steps with 1e-3.

**Comment by blisc** [comment]
The tacotron2 implementation in OpenSeq2Seq is very different than the tacotron2 implementation in NeMo. These models are not very comparable.

Regarding FP32 vs mixed precision, the default for tacotron2 should be FP32. There is an error in the documentation. I'll address it in the next update. Yes I would try FP32 as well as O1 to see if there is a difference.

**Comment by knbunmlr** [comment]
Good news: Training with O0 yields a stable training with learned attention already around 13k steps. Also the loss is not exploding anymore. 

One more thing where your expertise could help: The model seems to have problems to learn the stop gate, cf. the attached pictures. Any ideas how to improve this? Or is it possible that only the file shown in the tensorboard is problematic?
![image](https://user-images.githubusercontent.com/60105425/73343189-6e2e5a00-4280-11ea-9415-8e8f37f0791b.png)
![image](https://user-images.githubusercontent.com/60105425/73343148-58b93000-4280-11ea-9f5d-3a51310064f1.png)

**Comment by blisc** [comment]
Your eval gate will almost never match the true gate from the evaluation file. This is due to the fact that the length of generated audio from Tacotron 2 varies due to attention.

Your learned gate looks good as it seems to reach the top around the same time that attention has read all of the input text embeddings.

**Comment by Arjunprasaath** [comment]
Guys @DennisMuellerKenbun @blisc how to check the attention graph, i mean where do i find it or how do i plot it
